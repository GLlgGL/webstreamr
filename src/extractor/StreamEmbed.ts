import { Context, Format, Meta, UrlResult } from '../types';
import { Extractor } from './Extractor';

export class StreamEmbed extends Extractor {
  public readonly id = 'streamembed';

  public readonly label = 'StreamEmbed';

  public supports(_ctx: Context, url: URL): boolean {
    return null !== url.host.match(/bullstream|mp4player|watch\.gxplayer/);
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.href };

    const html = await this.fetcher.text(ctx, url, { headers });

    const video = JSON.parse((html.match(/video ?= ?(.*);/) as string[])[1] as string);

    return [
      {
        url: new URL(`/m3u8/${video.uid}/${video.md5}/master.txt?s=1&id=${video.id}&cache=${video.status}`, url.origin),
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          height: parseInt(JSON.parse(video.quality)[0]),
          title: decodeURIComponent(video.title),
        },
      },
    ];
  };
}
