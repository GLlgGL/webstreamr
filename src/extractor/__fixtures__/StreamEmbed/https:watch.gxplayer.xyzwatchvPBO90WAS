<html lang="en"><link type="text/css" rel="stylesheet" id="dark-mode-custom-link"><link type="text/css" rel="stylesheet" id="dark-mode-general-link"><style lang="en" type="text/css" id="dark-mode-custom-style"></style><style lang="en" type="text/css" id="dark-mode-native-style"></style><head><meta charset="utf-8"><title>Warning</title><style>body,html{width:100%;height:100%;margin:0;padding:0;background-color:#151515;overflow:hidden}body{text-align:center;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;color:#fff}#horizon{display:table;height:100%;position:static;width:100%}#sun{display:table-cell;vertical-align:middle;position:static;margin:0 auto}a,h1,p{margin:0 auto;max-width:80%}h1,p{display:block}h1{font-size:48px;margin-bottom:5px}p{font-size:24px;max-width:475px}a{color:#4bf;text-decoration:none}a[role=button]{margin-top:15px;display:inline-block;color:#fff;font-weight:700;padding:10px 20px;border-radius:5px;background:#4bf;text-shadow:none;font-size:16px}a[role=button]:active{background:#3795cc}.smaller p{font-size:18px}.smaller h1{font-size:42px}@media (max-width:499px){p{font-size:16px;line-height:1.4;max-width:290px}a[role=button]{font-size:16px;padding:7px 15px}.smaller p{font-size:15px}.smaller a[role=button]{font-size:14px}.smaller h1{font-size:36px}}@media (max-width:300px){h1{font-size:36px}p{font-size:14px;line-height:1.4;max-width:250px}a[role=button]{margin-top:10px;font-size:14px}.smaller p{font-size:12px}.smaller a[role=button]{font-size:12px}.smaller h1{font-size:28px}}</style> </head><body><div id="horizon"><div id="sun"><h1>Warning</h1><p>Video is not ready.</p> </div></div></body></html>