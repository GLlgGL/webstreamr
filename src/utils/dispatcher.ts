import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
