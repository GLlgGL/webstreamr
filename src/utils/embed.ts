import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
