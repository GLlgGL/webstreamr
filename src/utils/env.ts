import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
