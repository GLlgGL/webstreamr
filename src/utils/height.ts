import { Context } from '../types';
import { CustomRequestInit, Fetcher } from './Fetcher';

export const guessHeightFromPlaylist = async (ctx: Context, fetcher: Fetcher, url: URL, init?: CustomRequestInit): Promise<number | undefined> => {
  const m3u8Data = await fetcher.text(ctx, url, init);

  const heights = Array.from(m3u8Data.matchAll(/\d+x(\d+)|(\d+)p/g))
    .map(heightMatch => heightMatch[1] ?? heightMatch[2])
    .filter(height => height !== undefined)
    .map(height => parseInt(height));

  return heights.length ? Math.max(...heights) : /* istanbul ignore next */ undefined;
};
