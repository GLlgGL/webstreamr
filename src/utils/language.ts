import { CountryCode } from '../types';

const countryCodeMap: Record<CountryCode, { language: string; flag: string; iso639: string | undefined }> = {
  multi: { language: 'Multi', flag: '🌐', iso639: undefined },
  de: { language: 'German', flag: '🇩🇪', iso639: 'ger' },
  en: { language: 'English', flag: '🇺🇸', iso639: 'eng' },
  es: { language: 'Castilian Spanish', flag: '🇪🇸', iso639: 'spa' },
  fr: { language: 'French', flag: '🇫🇷', iso639: 'fra' },
  hi: { language: 'Hindi', flag: '🇮🇳', iso639: 'hin' },
  it: { language: 'Italian', flag: '🇮🇹', iso639: 'ita' },
  ja: { language: 'Japanese', flag: '🇯🇵', iso639: 'jpn' },
  ko: { language: 'Korean', flag: '🇰🇷', iso639: 'kor' },
  mx: { language: 'Latin American Spanish', flag: '🇲🇽', iso639: 'spa' },
};

export const languageFromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].language;
};

export const flagFromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].flag;
};

export const iso639FromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].iso639;
};

export const findCountryCodes = (value: string): CountryCode[] => {
  const countryCodes: CountryCode[] = [];

  for (const countryCode in countryCodeMap) {
    if (value.includes(countryCodeMap[countryCode as CountryCode]['language'])) {
      countryCodes.push(countryCode as CountryCode);
    }
  }

  return countryCodes;
};
