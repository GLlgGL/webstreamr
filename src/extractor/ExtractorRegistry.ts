import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
