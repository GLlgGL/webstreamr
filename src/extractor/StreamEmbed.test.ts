import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { StreamEmbed } from './StreamEmbed';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new StreamEmbed(new FetcherMock(`${__dirname}/__fixtures__/StreamEmbed`))]);

const ctx = createTestContext();

describe('StreamEmbed', () => {
  test('watch.gxplayer.xyz', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://watch.gxplayer.xyz/watch?v=MEKI92PU'))).toMatchSnapshot();
  });
});
