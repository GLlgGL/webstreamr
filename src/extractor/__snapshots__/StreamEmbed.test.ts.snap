// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StreamEmbed watch.gxplayer.xyz 1`] = `
[
  {
    "format": "hls",
    "label": "StreamEmbed",
    "meta": {
      "countryCodes": [],
      "height": 720,
      "title": "Baymax - Riesiges Robowabohu.mp4",
    },
    "sourceId": "streamembed_",
    "ttl": 900000,
    "url": "https://watch.gxplayer.xyz/m3u8/3/30b4a92517ace5825f5944c8a794ad3e/master.txt?s=1&id=11782&cache=1",
  },
]
`;
